File: spikes/native-midi-probe.mjs
// Spike B step 2 (OPTIONAL) — run ONLY if Chrome received no events while
// Serato was open. Diagnostic script run manually; nothing native ships.
//   npm install easymidi && node native-midi-probe.mjs   (with Serato open)
import easymidi from 'easymidi';
const inputs = easymidi.getInputs();
console.log('MIDI inputs visible natively:', inputs);
const name = inputs.find(n => /ddj|rev/i.test(n)) || inputs[0];
if (!name) { console.log('No inputs at all -> OS-level block or device asleep.'); process.exit(0); }
console.log(`Listening on "${name}" — touch the controller. Ctrl-C to stop.`);
const input = new easymidi.Input(name);
let count = 0;
for (const kind of ['noteon', 'noteoff', 'cc']) {
  input.on(kind, m => { count++; if (count % 25 === 1) console.log(kind, m); });
}
setInterval(() => console.log(`${count} events so far`), 2000);

File: spikes/native_midi_probe.py
# Spike B step 2 (OPTIONAL) - run ONLY if Chrome received no events while
# Serato was open. Diagnostic run manually; nothing native ships.
#   pip install mido python-rtmidi && python3 native_midi_probe.py
import mido, time
names = mido.get_input_names()
print("MIDI inputs visible natively:", names)
name = next((n for n in names if "DDJ" in n.upper() or "REV" in n.upper()), names[0] if names else None)
if not name:
    print("No inputs at all -> OS-level block or device asleep."); raise SystemExit
print(f'Listening on "{name}" - touch the controller. Ctrl-C to stop.')
count = 0
with mido.open_input(name) as port:
    last = time.time()
    for msg in port:
        count += 1
        if count % 25 == 1: print(msg)
        if time.time() - last > 2: print(count, "events so far"); last = time.time()
